### PR TITLE
Chemberta lm

### PR DESCRIPTION
**Pull request by seyonechithrananda**
# Pull Request Template

## Description

Add a wrapper for ChemBERTa's Masked language model.

Relies on ChemBERTa for base torch model with masked head, but wraps training + dataset functionalities into model class.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist
This is a WIP PR so styling, comments, etc have not been added yet.

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

The main thing I'm wondering about is whether I can overwrite TorchModel's `fit()` method as done so here, so I can pass the train and eval dataset to HuggingFace's trainer API (this is the basic flow of how ChemBERTa models, datasets and metrics are passed to the Trainer class).

@rbharath Would love your thoughts here! Its still not finished and need to tidy up a lot of the code but wanted to push this here to get your thoughts on the fit() method.

### CODE REVIEW

File: deepchem/models/torch_models/chemberta.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            num_train_epochs=num_train_epochs,
+            per_device_train_batch_size=per_device_train_batch_size,
+            save_steps=save_steps,
+            save_total_limit=save_total_limit,
+            fp16 = is_gpu and fp16, # fp16 only works on CUDA devices
+            report_to="wandb",
+            run_name=run_name,
+        )
+    
+    def _fit(self):
```
Comment (seyonechithrananda) [inline deepchem/models/torch_models/chemberta.py:160]: @rbharath Would love your thoughts on if it is ok to overwrite the TorchModel fit so we can fit the model into the Trainer object instead

File: deepchem/models/torch_models/chemberta.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
+"""
+Huggingface/transformers RoBERTa model for sequence-based property prediction.
+"""
+
+from numpy.lib.npyio import save
+import torch.nn as nn
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:6]: These should be wrapped in a try/catch block (since pytorch is an optional dependency)

File: deepchem/models/torch_models/chemberta.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import torch.nn.functional as F
+from torch.utils.data import random_split
+import os
+
+class ChemBERTaMaskedLM(nn.Module):
+    # defaults here match HF model: DeepChem/SmilesTokenizer_PubChem_1M.
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:12]: It looks like you're using 4-space spacing. Make sure to run `yapf` on these files and convert to DeepChem 2-space spacing

File: deepchem/models/torch_models/chemberta.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    # TODO - figure out which to use as default, and which to pass to model class
+    def __init__(self,
+                vocab_size =600,
+                max_position_embedddings = 515,
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:16]: We'll want to add type annotations here for these arguments

File: deepchem/models/torch_models/chemberta.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                num_hidden_layers = 6,
+                type_vocab_size = 1,
+                dataset_path: str = '',
+                mode = 'pre-trained',
+                model_path = 'DeepChem/SmilesTokenizer_PubChem_1M',
+                tokenizer_output_dir = 'tokenizer/',
+                tokenizer_type = 0, # if 0 - BPE, else if 1 - ST
+                max_tokenizer_len = 512,
+                BPE_min_frequency = 2,
+                **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:27]: We need a docstring here

File: deepchem/models/torch_models/chemberta.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                eval_steps: int = 1000,
+                logging_steps: int = 100,
+                overwrite_output_dir: boolean = True,
+                num_train_epochs: int = 10,
+                per_device_train_batch_size: int = 64,
+                save_steps :int = 10000,
+                save_total_limit: int = 2,
+                fp16: bool= True,
+                run_name: str = 'chemberta_lm'
+                **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:111]: Docstring here too

File: deepchem/models/torch_models/chemberta.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+class ChemBERTaMaskedLM(nn.Module):
+    # defaults here match HF model: DeepChem/SmilesTokenizer_PubChem_1M.
+    # TODO - figure out which to use as default, and which to pass to model class
+    def __init__(self,
+                vocab_size =600,
+                max_position_embedddings = 515,
+                number_attention_heads = 12,
+                num_hidden_layers = 6,
+                type_vocab_size = 1,
+                dataset_path: str = '',
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:20]: This design requires the model to have the dataset specified in the constructor which doesn't match the rest of DeepChem's design.

File: deepchem/models/torch_models/chemberta.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+        self.tokenizer = RobertaTokenizerFast.from_pretrained(tokenizer_path, max_tokenizer_len)
+
+class ChemBERTaMaskedLMModel(TorchModel):
+    def __init__(self,
+                vocab_size: int = 600,
+                max_position_embeddings = 515,
+                number_attention_heads=12,
+                num_hidden_layers=6,
+                dataset_path: str = '',
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:96]: Same issue here. The dataset should only be specified in the `fit()` method

File: deepchem/models/torch_models/chemberta.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            max_position_embeddings = max_position_embeddings,
+            number_attention_heads = number_attention_heads,
+            num_hidden_layers = num_hidden_layers,
+            dataset_path = dataset_path,
+            mode = mode,
+            model_path = model_path,
+            model_output_dir = model_output_dir,
+            **kwargs)
+                
+        dataset = LineByLineTextDataset(file_path=dataset_path,
```
Comment (rbharath) [inline deepchem/models/torch_models/chemberta.py:128]: We should see if we can use a DeepChem dataset instead. Also, the dataset shouldn't be constructed in the constructor

File: deepchem/models/torch_models/chemberta.py (hunk 1 above)
Comment (rbharath) [reply deepchem/models/torch_models/chemberta.py:160]: It's ok to overwrite if necessary but not if we can't satisfy the standard TorchModel API. I think write now the user would be surprised since they can't train this model with DeepChem datasets or standard deepchem flows

**Review by rbharath — COMMENTED** [review]
I did a first pass through. The major work left to do is smoothing the interaction with the DeepChem API here. I've left some comments below with suggestions on how to start :)

File: deepchem/models/torch_models/chemberta.py (hunk 9 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py:128]: Sounds good - I'll try playing around with diskdataset and see if I can get it working! I'll also move over the dataset code into _fit()

Comment (vsomnath) [reply deepchem/models/torch_models/chemberta.py:128]: How important is the reliance on the `Trainer` class? If parts of the code can be used from `Trainer` (like DataLoader creation moved to `default_generator`, and we have a similar loop in `fit_generator`, then that would simplify things.

File: deepchem/models/torch_models/chemberta.py (hunk 2 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py:7]: Will do - nice catch!

File: deepchem/models/torch_models/chemberta.py (hunk 3 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py:16]: yup! right now i'm just working on getting all the components to fit better with the deepchem API + pass the unit test I'm using - still have to fix formatting + commenting :)

File: deepchem/models/torch_models/chemberta.py (hunk 4 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py:20]: good catch! will do

File: deepchem/models/torch_models/chemberta.py (hunk 7 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py:24]: Good point! For now, I've moved the dataset initialization, data collator, + training args to the `_fit()` method, is that ok? 

This would be until I can get Disk Dataset working of course

File: deepchem/models/torch_models/chemberta.py (hunk 8 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py:99]: Done!

File: deepchem/models/torch_models/chemberta.py (hunk 9 above)
Comment (seyonechithrananda) [reply deepchem/models/torch_models/chemberta.py]: The [trainer class](https://huggingface.co/transformers/_modules/transformers/trainer.html#Trainer) might be a considerable amount of technical debt to inherit + maintain - mainly because all of the Huggingface model infra revolve around passing into a trainer object. We use the trainer class to pass shards from the dataset into the transformer, enable mixed precision (fp16) training as done in the chemberta paper, compute masked loss during training and push to the Deepchem model hub.  I can try to use pure PyTorch code instead, but I think it may be a lot less stable and easier to maintain. 

What do you guys think? If I can try to get an end-to-end Chemberta deepchem integration going (i.e - feature done, models for pre-training + fine-tuning, DiskDataset integration, etc), we can go back and improve the compatibility if that works!

Comment (rbharath) [reply deepchem/models/torch_models/chemberta.py]: I think using the trainer class under the hood seems fine so long as we can have the external API of the class match DeepChem's usual API. Is the code in this PR ready for a second round of review?
